Description: Fix a memory leak with invalid input.
Forwarded: https://github.com/rpm-software-management/drpm/pull/6
Author: Peter Pentchev <roam@ringlet.net>
Last-Update: 2020-06-01

--- a/src/drpm_decompstrm.c
+++ b/src/drpm_decompstrm.c
@@ -746,14 +746,14 @@
             return DRPM_ERR_IO;
     }
 
+    if (in_len == 0)
+        return DRPM_ERR_FORMAT;
+
     size_t const buffOutSize = ZSTD_DStreamOutSize();
     void* const buffOut = malloc(buffOutSize);
     if (buffOut == NULL)
         return DRPM_ERR_MEMORY;
 
-    if (in_len == 0)
-        return DRPM_ERR_FORMAT;
-
     ZSTD_inBuffer input = { in_buffer, in_len, 0 };
 
     while (input.pos < input.size) {
